### ISSUE TITLE
tuple unpacking in `__init__` causes erroneous "Type is missing type annotation"

### ISSUE THREAD

**Issue by cjw296** [microsoft/pyright#8957]
**Describe the bug**
See the change in this pull request: https://github.com/simplistix/sybil/pull/124

It would appear that using tuple unpacking in a constructor triggers a bug in pyright that causes it to fail to correctly infer the types of the assigned attributes.

**Code or Screenshots**

This code should not trigger "Type is missing type annotation":

```python
class Lexeme:
    def __init__(self, text: str, offset: int, line_offset: int) -> None:
        self.text, self.offset, self.line_offset = text, offset, line_offset
```

The workaround is to remove the tuple unpacking, which should not be necessary:

```python
class Lexeme:
    def __init__(self, text: str, offset: int, line_offset: int) -> None:
        self.text, self.offset, self.line_offset = text, offset, line_offset
```

```python
class Lexeme:
    def __init__(self, text: str, offset: int, line_offset: int) -> None:
        self.text = text
        self.offset = offset
        self.line_offset = line_offset
```

**Comment by erictraut** [comment]
Pyright is able to infer the types of these instance variables without a problem. However, other type checkers and language servers will not. For that reason the `--verifytypes` command reports this as an error.

If you want your code to be compatible with static type analyzers in general, you have a couple of options:

1. Don't use tuple unpacking to initialize instance variables in your `__init__` methods. Move each assignment to its own line.
2. Provide explicit type annotations for the instance variables, like this:
```python
class Lexeme:
    text: str
    offset: int
    line_offset: int

    def __init__(self, text: str, offset: int, line_offset: int) -> None:
        self.text, self.offset, self.line_offset = text, offset, line_offset
```

**Comment by cjw296** [comment]
> Pyright is able to infer the types of these instance variables without a problem.

So is mypy, so what are the other type checkers and language servers you're referring to?

Given the two predominant static type analyzers for python are compatible with tuple unpacking in assignment, what is the basis for the assertion that, in general, static type analyzers will not be able to do this?

**Comment by erictraut** [comment]
The [Python type spec](https://typing.readthedocs.io/en/latest/spec/index.html) is the official document that specifies standardized type checker behaviors. While it mandates many behaviors, it doesn't provide any guidance or specifications regarding inference of instance variable types. Maybe it should, but I think it will be difficult to get all type checker authors on the same page with regard to inference behaviors.

The current behavior of pyright's `--verifytypes` tool was informed by discussions with other library and tools authors. For example, see [this discussion](https://github.com/python/typing/discussions/1058) and [this discussion](https://github.com/python/typing/pull/913#issuecomment-962098501).

If you want your library to work with static analysis tools across the ecosystem (type checkers, linters, language servers, etc.), then I recommend following [this guidance](https://microsoft.github.io/pyright/#/typed-libraries?id=type-completeness). The `--verifytypes` tool is designed to help library authors achieve this goal. Of course, as a library author, you're free to ignore any of its suggestions.
